engine/common/src/main/java/org/enso/common/CachePreferences.java

```diff
@@ -0,0 +1,98 @@
+package org.enso.common;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
+
+/**
+ * Contains settings telling which nodes are marked for caching and their kinds.
+ *
+ * @param preferences the config with cached preferences
+ */
+public record CachePreferences(Map<UUID, Kind> preferences) {
+
+  /**
+   * @return empty cache preferences config
+   */
+  public static CachePreferences empty() {
+    return new CachePreferences(new HashMap<>());
+  }
+
+  /** A kind of cached value. */
+  public enum Kind {
+    BINDING_EXPRESSION,
+    SELF_ARGUMENT
+  }
+
+  /**
+   * Get the cache preference for the provided node id.
+   *
+   * @param id the node id
+   * @return the kind of cached value if available
+   */
+  public Kind get(UUID id) {
+    return preferences.get(id);
+  }
+
+  /**
+   * Get all the node ids marked for caching for a particular node kind.
+   *
+   * @param kind the node kind
+   * @return the set of node ids marked for caching
+   */
+  public Set<UUID> get(Kind kind) {
+    var result = new HashSet<UUID>();
+    for (var entry : preferences.entrySet()) {
+      if (entry.getValue().equals(kind)) {
+        result.add(entry.getKey());
+      }
+    }
+
+    return result;
+  }
+
+  /**
+   * Add a cache preference.
+   *
+   * @param id the node id
+   * @param kind the node kind
+   */
+  public void set(UUID id, Kind kind) {
+    preferences.put(id, kind);
```

Comment:
I am afraid of a data structure read from _persistance_ that can be modified. At certain moment such a structure must become immutable. We cannot have a mutable structures flying around in the system. There are already plenty of such structures and we don't want to introduce new ones.

Reply:
> I am afraid of a data structure read from persistance that can be modified.

From what I can see, all the compiler metadata is mutable currently. I introduced the `CachePreferences` to avoid copying and simply pass it to runtime after the compilation.

If we want to introduce an immutable object for persistence, we would need to convert mutable prefs to immutable to pass it to the persistence. And then convert immutable prefs to mutable again when we pass it to the runtime.